[WPE][GTK] Speech synthesis using libspiel

philn opened the pull request:
#### 290d009f7b35d6af9bf5b5b80d113f5db6a04274
<pre>
[WPE][GTK] Speech synthesis using libspiel
<a href="https://bugs.webkit.org/show_bug.cgi?id=270742">https://bugs.webkit.org/show_bug.cgi?id=270742</a>

Reviewed by Xabier Rodriguez-Calvar.

Spiel&apos;s API is much nicer than Flite&apos;s, being directly inspired from WebSpeech. So the speech
synthesis is now handled with Spiel in the WPE and GTK ports. The feature is also enabled by
default. A couple layout tests relying on the presence of voice providers in the SDK had to be
flagged, because the SDK doesn&apos;t ship any voice provider yet.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp: Removed.
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp: Removed.
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h: Removed.
* Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp: Added.
(WebCore::SpielSpeechWrapper::SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::~SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::initializeVoiceList):
(WebCore::SpielSpeechWrapper::pause):
(WebCore::SpielSpeechWrapper::resume):
(WebCore::SpielSpeechWrapper::speakUtterance):
(WebCore::SpielSpeechWrapper::cancel):
(WebCore::SpielSpeechWrapper::resetState):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/cmake/FindLibSpiel.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/284276@main">https://commits.webkit.org/284276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85738acfb66c4ff558e64f9badba7f425c270b74

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21615 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73023 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20094 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19949 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13360 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72009 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59525 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35383 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16954 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18475 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62059 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17300 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74730 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68189 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12925 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16545 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62546 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59608 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62448 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15297 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10424 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4033 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89969 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44147 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15950 "Found 41 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Number/property_and_index_of_number.js.default, ChakraCore.yaml/ChakraCore/test/Strings/property_and_index_of_string.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/match_global.js.default, jsc-layout-tests.yaml/js/script-tests/dfg-convert-this-object-then-exit-on-other.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-dead-speculation.js.layout, stress/array-prototype-splice-making-typed-array.js.dfg-eager, stress/big-int-bitwise-and-type-error.js.default, stress/big-int-exp-type-error.js.mini-mode, stress/big-int-literals.js.bytecode-cache, stress/big-int-prototype-to-string-cast-overflow.js.bytecode-cache ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45220 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44962 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/FindLibSpiel.cmake:
```diff
@@ -0,0 +1,55 @@
+    NAMES spiel-1.0
+    HINTS ${PC_SPIEL_LIBDIR}
+          ${PC_SPIEL_LIBRARY_DIRS}
+)
```
Please change this to use imported targets. You can check e.g. `FindLibPSL.cmake` for inspiration on how to do that.

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,258 @@
+    // Async API can't really be used here because the voices initialization
+    // (PlatformSpeechSynthesizer::initializeVoiceList()) is synchronous.
+    GUniqueOutPtr<GError> error;
+    m_speaker = adoptGRef(spiel_speaker_new_sync(nullptr, &error.outPtr()));
```
A common pattern is to initialize speech in an async fashion and return an empty list for voices. Once it is initialized you call `voicesDidChange` to notify content that voices are available.

philn replied:
Yeah I tried that but it didn't work :( I think it needs additional work at the cross-platform level.

philn replied:
The demo I tried this with is https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/

eeejay replied:
That is strange since the [script.js](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/script.js) does listen voice "voiceschanged" to re-populate the dropdown. I'm curious if this indeed does get emitted for you. Maybe getVoices is retuning null instead of an empty list so the script fails?

eeejay replied:
That is strange since the [script.js](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/script.js) does listen voice "voiceschanged" to re-populate the dropdown. I'm curious if this indeed does get emitted for you. Maybe getVoices is retuning null instead of an empty list so the script fails?

eeejay replied:
This is what I see in a first time run of Chrome:
![image](https://github.com/user-attachments/assets/039f314c-924f-44df-8780-613c76cfea57)

This is the URL:
`data:text/html,<script>function populateVoices() {console.log('Voice count', speechSynthesis.getVoices().length);}populateVoices();speechSynthesis.onvoiceschanged = populateVoices;</script>`

philn replied:
Got it working with async :)

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,258 @@
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::voices", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker*) {
+        self->m_platformSynthesizer.client().voicesDidChange();
```
Maybe this happens somewhere else, but does the new voice list get reflected to content?

philn replied:
I'm not sure what you're asking :) `voicesDidChange()` triggers a `voiceschangedEvent`  (see `SpeechSynthesis::voicesDidChange()`) and a re-initialization of the voices list the next time they are queried from JS.

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,258 @@
+        if (!G_N_ELEMENTS(languages))
+            continue;
+
+        auto language = makeString(span(languages[0]));
```
Ideally, you would make a web voice for every language in a spiel voice. If you don't then the web will only see a single Afrikans eSpeak voice, for example. Instead of the ~130 other languages

philn replied:
Ah yes, good point!

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,258 @@
+
+        auto language = makeString(span(languages[0]));
+        auto name = makeString(span(spiel_voice_get_name(voice)));
+        auto identifier = makeString(span(spiel_voice_get_identifier(voice)));
```
The voice identifier is only unique to the provider, so theoretically you can have two separate providers using the same voice IDs for their voices. I would prefix the ID with the provider's well known name. Along with the multi-language voice support you would need as the suffix.

You can see how I did this for Firefox in `SpielService::GetVoiceURI` [here](https://phabricator.services.mozilla.com/D217313). (Firefox's speech backend API is very different though in many other regards)

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,258 @@
+    }
+    spiel_utterance_set_volume(spielUtterance.get(), utterance->volume());
+    spiel_utterance_set_pitch(spielUtterance.get(), utterance->pitch());
+    spiel_utterance_set_rate(spielUtterance.get(), utterance->rate());
```
Maybe not in an initial patch, but you can try to detect if the string is XML, and if it is call `spiel_utterance_set_ssml`

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -141,7 +141,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_THUNDER PRIVATE ${ENABLE_DEVELOPER_MODE}
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_POINTER_LOCK PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SHAREABLE_RESOURCE PRIVATE ON)
-WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPEECH_SYNTHESIS PRIVATE ${ENABLE_EXPERIMENTAL_FEATURES})
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPEECH_SYNTHESIS PRIVATE ON)
```
Please make it `PUBLIC` in OptionsGTK.cmake and OptionsWPE.cmake so distros can see it as an option.

eeejay commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,293 @@
+    spiel_utterance_set_pitch(spielUtterance.get(), utterance->pitch());
+    spiel_utterance_set_rate(spielUtterance.get(), utterance->rate());
+    m_utterance = WTFMove(utterance);
+    spiel_speaker_speak(m_speaker.get(), spielUtterance.get());
```
Is there a chance m_speaker is null here if speak is called before the async init is done? Looking at this on my phone so maybe I missed something :)

If init isn't done the speak() needs to be deferred.

philn replied:
Yep, I believe this is handled, scroll down a bit to `PlatformSpeechSynthesizer::speak()` ;)

eeejay replied:
😎 Thanks for working on this!

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> initializeVoiceList();
+    void pause();
+    void resume();
+    void speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&&);
```
can we has `setSpeakUtterance` ?

philn replied:
`speakUtterance` refers to the action of speaking an utterance. And it's consistent with the cocoa backend. I'd rather keep this name.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+        self->m_speaker = adoptGRef(spiel_speaker_new_finish(result, &error.outPtr()));
+        if (error) {
+            WTFLogAlways("Spiel speaker failed to initialize: %s", error->message);
+            return;
```
I think it would be a good idea to GST_ERROR as well and I would advice to ASSERT_NOT_REACHED and also setting `m_speaker` to null.

philn replied:
`spiel_speaker_new_finish()` already returns NULL on error.

philn replied:
I don't think asserting there is a good idea. iirc that can happen when the host has no voice provider, which is not a programmer error.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+        return;
+    }
+
+    spiel_speaker_new(nullptr, [](GObject*, GAsyncResult* result, gpointer userData) {
```
I'm curious of why you use the async API instead of the sync one.

philn replied:
sync dbus calls should be avoided when possible. Our main loop is busy enough already.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+
+String SpielSpeechWrapper::generateVoiceURI(const GRefPtr<SpielVoice>& voice, const String& language)
+{
+    auto provider = spiel_voice_get_provider(voice.get());
```
adopt, this is transfer full

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+    auto voices = spiel_speaker_get_voices(m_speaker.get());
+    unsigned position = 0;
+    m_voices.clear();
+    while (auto item = g_list_model_get_item(voices, position)) {
```
Some day we'll create an iterator for this kind of data structures.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+            platformVoices.append(PlatformSpeechSynthesisVoice::create(uri, name, language, false, isDefault));
+            m_voices.add(uri, GRefPtr(voice));
+        }
+        position++;
```
Move this `position++` into the while clause above. It is not used anywhere else.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+    return makeString(URI_PREFIX, span(spiel_provider_get_well_known_name(provider)), '#', span(spiel_voice_get_identifier(voice.get())), '#', language);
+}
+
+Vector<RefPtr<PlatformSpeechSynthesisVoice>> SpielSpeechWrapper::initializeVoiceList()
```
I don't know if it makes sense or not but I think you could split this into two methods:

* private initialize that stores the voices in an attribute.
* `voiceList` that returns `const Vector...  &` of that attribute.

philn replied:
I don't think that's needed :)

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+    ASSERT(!m_utterance);
+    ASSERT(utterance);
+    if (!utterance)
+        return;
```
Use ASSERT_NOT_REACHED inside the if and maybe a GST_ERROR.

philn replied:
What is the point of adding another assert? We have one above already.

calvaris replied:
The point would be removing the one above because you're checking the condition twice.

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    const auto& uri = utterance->voice()->voiceURI();
+    if (!m_voices.contains(uri)) {
+        WTFLogAlways("Unknown voice URI: %s", uri.utf8().data());
```
GST_ERROR and maybe ASSERT_NOT_REACHED

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,294 @@
+    auto voice = m_voices.get(uri);
+    auto text = utterance->text().utf8();
+    auto spielUtterance = adoptGRef(spiel_utterance_new(text.data()));
+    auto language = utterance->lang().utf8();
```
I would avoid `text` and `language` variables. Keep them if you feel like it but at least keep them as a String, the byproduct of calling `utf8()` gives me the shivers.

philn committed:
[WPE][GTK] Speech synthesis using libspiel
https://bugs.webkit.org/show_bug.cgi?id=270742

Reviewed by Xabier Rodriguez-Calvar.

Spiel's API is much nicer than Flite's, being directly inspired from WebSpeech. So the speech
synthesis is now handled with Spiel in the WPE and GTK ports. The feature is also enabled by
default. A couple layout tests relying on the presence of voice providers in the SDK had to be
flagged, because the SDK doesn't ship any voice provider yet.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp: Removed.
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp: Removed.
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h: Removed.
* Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp: Added.
(WebCore::SpielSpeechWrapper::SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::~SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::initializeVoiceList):
(WebCore::SpielSpeechWrapper::pause):
(WebCore::SpielSpeechWrapper::resume):
(WebCore::SpielSpeechWrapper::speakUtterance):
(WebCore::SpielSpeechWrapper::cancel):
(WebCore::SpielSpeechWrapper::resetState):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/cmake/FindLibSpiel.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/284276@main
